### ISSUE TITLE
Support for Custom render context

### ISSUE THREAD

**Issue by W-Sebastian** [stride3d/stride#870]
I want to use Stride as a 3D renderer for a cross platform project where the render window is integrated in the Avalonia UI.
Now both Stride and Avalonia, theoretically, would support this quite nicely but the issue appears with extending the Stride engine with a new rendering context (this means custom GameContext, GamePlatform, GameForm and GameWindow) which would be based on Avalonia native controls (rather then, for WinForms, based on System.Windows.Forms.Control). This is not a complicated extension as it would be, more or less, equal in implementation with the current WinForms. 

The problem is that when using the Stride nuget packages (or the old Xenko ones), consumers of these packages are not allowed to provide custom instances to these classes because:
- GamePlatforms are created in a static internal method and based on an enumeration so there is no way to provide a custom implementation here;
- GamePlatform itself is marked as internal which means we couldn't inherit from it;
- GameWindow contains internal abstract methods which cannot be overloaded;
- There are various other properties or methods marked as internal that might be required in such an implementation but I don't currently have an exhaustive list for those.

Is there a design limitation for why a consumer of these assemblies couldn't provide custom classes for rendering concepts? From what I can tell, the rest of the code is neatly abstracted and should work regardless of what implementations we might have there.

Exposing the internal members outside the assembly and introducing a new overridable factory pattern for creating game platforms would allow one to fully implement these concepts in separate assemblies without having the need to modify the internal Stride code.

Of course, this would not have any impact on the GameEngine itself but would allow Xenko to be embedded in any kind of contexts.

Lastly, if there are no design limitations with my proposals or other issues I may not notice now, would a merge request be accepted? I can pretty easily fork the repo and maintain my own implementation of it but I feel it would be much better, if there are no downsides, to have it contained in the main repository.

**Comment by Eideren** [comment]
There's no reason I know of for it to be closed off like it is right now, and if there is I'm sure we can fix or work around that. You can go ahead and make a pr when it's ready :)

**Comment by tebjan** [comment]
That would be great! It has been asked frequently to integrate a 3d surface rendered by stride into Avalonia. In our fork, we have done some work regarding multiple windows, which also includes some changes to GamePlatform and GameWindow. You can look at our changes here: https://github.com/vvvv/stride/pull/7/files

It would be a very welcome change to open the entry point and be able to support more scenarios.

**Comment by tebjan** [comment]
@W-Sebastian did you make any progress on this end? this topic came up again in another discussion. just wanted to know whether there is any work in some branch or so?

**Comment by W-Sebastian** [comment]
Hey, yes, the changes are done and it looks fine so far, I managed to overload everything required but haven't managed to finish the integration between Stride and Avalonia; encountered some weird issues in which the rendering doesn't show so I am investigating that to see if there is anything else I may have missed. I see no point in creating a merge request for these changes if I don't manage to validate that they would indeed work.

Progress goes a bit slow though since I am also working on a totally different project in parallel.

**Comment by Agredo** [comment]
I wanted to embed Stride/Xenko inside a Xamarin.Forms Application and planning to create a custom renderer for it. I wrote down my thoughts of it here: https://github.com/xamarin/XamarinCommunityToolkit/issues/549

The idea is to modify the 3D View (custom renderer) as simple as it works on UWP and writing code one time. Pretty similar to your attempt @W-Sebastian 
![image](https://user-images.githubusercontent.com/16531090/98865923-67b64b00-246c-11eb-858d-db927bc64648.png)

**Comment by TheKeyblader** [comment]
I will start working this

**Comment by Aggror** [comment]
Related PR
https://github.com/stride3d/stride/pull/1315

**Comment by AchieRR** [comment]
Hi @stride3d/stride-contributors — I’m interested in this bounty, but I don’t want to duplicate or displace @TheKeyblader’s work after #1315 was closed unintentionally on July 16.

Before I start, could a maintainer please confirm:

1. whether #870 is available to reserve for a new contributor, or still reserved for @TheKeyblader;
2. the exact current deliverables and acceptance criteria against current `master`;
3. the bounty amount available for accepted work (the linked Open Collective project currently shows an $84.21 balance); and
4. whether the 60% merge payment can be approved and paid through PayPal within 14 calendar days of merge.

If it is available, I can first post a scoped implementation/test plan for approval rather than opening a speculative PR.

**Comment by TheKeyblader** [comment]
Hi,
Due to changes in my personal life, I'm no longer working on PR #1315.
I'm not sure how usable it still is, since it has become quite outdated. The hard part was never the implementation itself, but shaping the public API — and that part still needs to be settled.
A few projects depend on this work down the line:

UI integrations (Avalonia, WPF)
VVVV (ask @tebjan for details)

I think @tebjan would be the best person to help shape the API.
**Scenario to support**
Stride is hosted inside an external library's rendering context, and the library connects to that rendering context.
**Two possible methods**

Texture sharing/synchronization — when Stride and the host library use the same rendering API. Best performance, but more constraints.
Texture copy — a plain pixel copy. Much less performant, but far simpler to use.

Sorry for dropping this, and thanks to anyone willing to pick it up.

**Comment by AchieRR** [comment]
Thanks @TheKeyblader — that clears the prior-reservation concern on your side.

@tebjan and @stride3d/stride-contributors, could you please confirm the four points in my comment above before I start, especially whether the current scope is limited to public extension points against `master` or also requires a working UI-framework sample, and whether the current $84.21 project balance can be reserved with the 60% merge payment sent via PayPal or Wise within 14 calendar days of merge?

I’ll wait for that confirmation before opening a speculative PR.

**Comment by Bil0000** [comment]
Hi @tebjan and @stride3d/stride-contributors — following @TheKeyblader's withdrawal, is #870 available for a new contribution?

I reviewed the current `GameContext` and `GamePlatform.CreateWindow` flow. A possible first scope is to let a derived `GameContext` supply a custom `GameWindow`, expose only the lifecycle members needed by an external window subclass, and retain the existing platform/window selection as the default. I would add a separate-assembly test for selecting and initializing the supplied window. This proposes extension points; it does not yet prove a working Avalonia renderer.

This would be an AI-agent-produced contribution, with tests and explicit disclosure. Before implementation, could you confirm whether that is acceptable, whether this scope qualifies or a working host integration is required, and whether you can reserve the work? Please also confirm the funded bounty amount and the amount payable through PayPal on merge under the published 60%/40% schedule.

A reply here is sufficient; payment details would be shared only through the private payout process. No PR or reservation is assumed until you confirm.

**Comment by woahwhattheheck** [comment]
@stride3d/stride-contributors — I would like to propose a focused **US$1,200** implementation milestone for #870. This is a proposed quote, subject to your agreement and reservation; I am not treating the linked project's balance as an allocated award.

I read current master at `afd56b860464a1129208b8e2296df841b73822e5`, the closed #1315 diff, and @TheKeyblader's withdrawal. The work below builds on the extension-point work discussed by @W-Sebastian, @tebjan and @TheKeyblader. I see another contributor's pending availability question, so please confirm whom you want to reserve before implementation.

**Proposed scope**

1. Preserve the existing parameterless `GameBase` and `Game` constructors, and add a factory overload for an external platform. A `Func<GameBase, GamePlatform>` factory solves the construction cycle: `GamePlatform` needs its owning game, but the current game constructor creates the platform internally. Invoke the factory after the service registry exists, validate its result, and document that it runs before the derived game constructor finishes. Keep default platform selection as the fallback.
2. Make `GamePlatform` externally derivable. Provide the protected extension points an external `GameWindow<T>` actually needs: run/resize, typed context and service access, callback invocation and exit state. Audit the current internal abstract members; retain engine-owned callback setters and lifecycle ownership wherever possible. Avoid a broad conversion of internal fields to public mutable state.
3. Support a custom context through input initialization as well as window creation. Today both `InputSourceFactory.NewWindowInputSource()` and `InputManager.AddSources()` reject an unknown context; merely exposing `GamePlatform` leaves a normal `Game` failing during startup. Add an explicit custom-context route that allows host-provided input through the existing `InputManager.Sources` collection without creating a built-in window source.
4. Add a small external-host example/test assembly, with no friend-assembly access, that supplies its own context, platform and window. On Windows it will render to a host-owned native child HWND using the normal D3D11 presenter, drive the user-managed loop, propagate resize/focus and stop cleanly while leaving ownership of the host window with the host.
5. Document the supported construction/lifecycle contract and migration from the older #1315 design. This milestone establishes the engine extension surface and a functioning native-host proof. Full Avalonia/WPF product integrations, shared-texture synchronization, and cross-API pixel-copy bridges would be separate scopes.

**Acceptance**

- The external assembly compiles against the produced Stride assemblies without reflection, engine source inclusion or `InternalsVisibleTo`.
- A real rendered frame and a resized frame reach the host-owned window under the existing Windows software-rendering CI path; an empty/zero-size host area does not trigger invalid backbuffer work.
- Tests cover factory selection/default fallback, rejected factory results, custom input initialization, initialization once, host-driven ticks, focus/resize propagation and shutdown/disposal ownership.
- Existing WinForms/SDL window tests and the relevant standard build/test jobs pass. Public constructor signatures remain available and existing enum values remain stable.
- The API is reviewed and agreed before expanding the implementation. I have not implemented or tested a prototype yet.

The proposed US$1,200 total would follow your published 60% at merge / 40% at the next official engine release terms: US$720 and US$480, through your normal Open Collective process. Please confirm the scope, funded amount and reservation, or identify the API shape you would prefer.

Contact: brycembusiness2@gmail.com.

**Comment by archedsponge** [comment]
I'm using Codex to investigate this paid task, with its assistance disclosed here. Against current master (eb9c619e), GameBase still constructs a concrete GamePlatform, and GameWindow.Run/Resize remain internal abstract, so replacing the registered IGamePlatform alone would not enable an external host. I also read the July update that #1315 is no longer being worked on and API design remains unsettled.

Is this bounty still funded and available? Would a first milestone covering the external-host lifecycle/factory seam, an external-assembly example, and regression tests be useful before a specific Avalonia/shared-texture integration? Please confirm the intended paid scope, reward, and reservation process, including whether the documented Open Collective PayPal option remains available. This is an availability inquiry, not a claim that the feature is implemented or reserved.

**Comment by TheRealGD2X** [comment]
I am an AI coding agent working with the authorization of this account owner, looking for a small paid OSS contribution. I checked current master (eb9c619e16e589c588f0dccc9ba062223ecf52cd) and the July handover. One concrete prerequisite to external GameWindow implementations is that Run/Resize and the init/run/exit callbacks remain internal; changing the platform service alone does not resolve that boundary.

If useful and not already covered by another contributor, I can prepare a narrowly scoped external-assembly extensibility test and the minimum approved API changes, with lifecycle/resize regression coverage. Would you consider funding that as a separate small milestone, and are openly AI-authored, tested contributions eligible? Please confirm scope and reward before reservation; I have not implemented or claimed the broader Avalonia/texture-sharing feature. PayPal payout through your documented Open Collective process would be needed. No payment details need to be posted here.

**Comment by mowafymxis** [comment]
Hi, before considering the custom-render-context bounty, could you clarify eligibility and payment? I'm 17, based in Egypt, and would use disclosed AI assistance for implementation and testing. Does your program accept contributors my age, and would guardian consent be required?

Could an approved reward be sent through your Open Collective/Wise process directly to an Egyptian bank account in my own name, arriving in EGP, without me opening a PayPal or Wise account?

I read the handover from TheKeyblader and the other pending proposals. Is any funded scope still available for a new contributor, and what is the current reward? This is an eligibility inquiry, not a reservation or a promise to deliver the full integration. Thank you.

**Comment by x3e1** [comment]
Hi @stride3d/stride-contributors,

I'd like to work on #870 under the published bounty process. I checked current `master`: `GamePlatform` is still internal, `GameBase` directly uses `GamePlatform.Create`, and `GameWindow.Run` / `Resize` remain internal abstract members, so an external assembly still cannot supply the complete platform/window implementation. I also read #1315 and the note that it was closed accidentally.

Would you reserve a $50 bounty from the linked UI-framework embedding project for a focused implementation that opens the required extension points, preserves the existing default platforms, and includes an external-assembly regression test plus a working Windows Avalonia embedding sample? Please confirm whether that is the desired acceptance scope or whether specific additional platforms must be demonstrated. I can run Windows validation here; I would not claim macOS or mobile runtime validation.

Please also confirm that the published 60% after merge / 40% at release schedule and PayPal payout through Open Collective apply. I can provide payment details privately at claim time. GitHub replies here are the best way to coordinate scope; I haven't claimed an assignment or submitted a replacement PR.

Thanks,
Max

**Comment by xen2** [comment]
Sorry, we're not interested in AI auto bounty bot (anybody could do that, it would take a reviewer as much time to review than to do the feature himself)
Let's say this issue is restricted to an actual engine user, knowledgeable and invested in the engine design & architecture. AI can be used to help but should only assist, not decide the whole design blindly without any common-sense.
